src/cache/redis.rs

```diff
@@ -2,37 +2,95 @@ use crate::twin::Twin;
 
 use super::Cache;
 
-use anyhow::Result;
+use anyhow::{Context, Result};
 use async_trait::async_trait;
-use bb8_redis::{bb8::Pool, RedisConnectionManager};
-use serde::{Deserialize, Serialize};
+use bb8_redis::{
+    bb8::{Pool, PooledConnection},
+    redis::{cmd, FromRedisValue, ToRedisArgs},
+    RedisConnectionManager,
+};
+use serde::{de::DeserializeOwned, Serialize, Serializer};
 
 //
 // how_to_init
 // let manager = RedisConnectionManager::new(url)?;
 // let pool = Pool::builder().build(manager).await?;
+// RedisCache::new(pool, "twin", 10); //10 seconds
 //
 
 #[derive(Clone)]
-pub struct RedisCache {
+pub struct RedisCache<P>
+where
+    P: Into<String>,
+{
     pool: Pool<RedisConnectionManager>,
+    prefix: P,
+    ttl: usize,
 }
 
-impl RedisCache {
-    pub async fn new(pool: Pool<RedisConnectionManager>) -> Result<Self> {
-        Ok(Self { pool })
+impl<P> RedisCache<P>
+where
+    P: Into<String>,
+{
+    pub async fn new(
+        pool: Pool<RedisConnectionManager>,
+        prefix: P,
+        ttl_sec: usize,
+    ) -> Result<Self> {
+        Ok(Self {
+            pool,
+            prefix,
+            ttl: ttl_sec,
+        })
+    }
+
+    async fn get_connection(&self) -> Result<PooledConnection<'_, RedisConnectionManager>> {
+        let conn = self
+            .pool
+            .get()
+            .await
+            .context("unable to retrieve a redis connection from the pool")?;
+
+        Ok(conn)
     }
 }
 
 #[async_trait]
-impl<'a, T> Cache<T> for RedisCache
+impl<T, P> Cache<T> for RedisCache<P>
 where
-    T: Serialize + Deserialize<'a> + Send + Sync + 'static,
+    T: Serialize + DeserializeOwned + Send + Sync + 'static,
+    P: Into<String> + Send + Sync + Clone,
 {
-    async fn set<S: ToString + Send + Sync>(&self, id: S, obj: T) -> Result<()> {
+    async fn set<S: ToString + Send + Sync>(&self, key: S, obj: T) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+        let obj = serde_json::to_vec(&obj).context("unable to serialize twin object for redis")?;
+        let mut key = format!("{:?}.{}", self.prefix.clone().into(), key.to_string());
```

Comment:
Once the type is fixed, this can be easily written as 

```suggestion
        let key = format!("{}.{}"), self.prefix, key.to_string());
```